File: leetCode/Python/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # two pointer
        # res = []
        # length = len(nums)
        # if not nums or length < 3:
        #     return res
        # nums.sort()
        # for i in range(length):
        #     if nums[i] > 0:
        #         return res
        #     if (i > 0 and nums[i] == nums[i - 1]):
        #         continue
        #     left = i + 1
        #     right = length - 1
        #     while left < right:
        #         if nums[i] + nums[left] + nums[right] == 0:
        #             res.append([nums[i], nums[left], nums[right]])
        #             while (left < right and nums[left] == nums[left+1]):
        #                 left += 1
        #             while (left < right and nums[right] == nums[right-1]):
        #                 right -= 1
        #             right -= 1
        #         elif nums[i] + nums[left] + nums[right] > 0:
        #             right -= 1
        #         else:
        #             left += 1
        # return res
                    
        # hashset
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i == 0 or nums[i - 1] != nums[i]:
                self.twoSum(nums, i, res)
        return res
        
    def twoSum(self, nums: List[int], i: int, res: List[List[int]]):
        seen = set()
        j = i + 1
        while j < len(nums):
            complement = -nums[i] - nums[j]
            if complement in seen:
                res.append([nums[i], nums[j], complement])
                while j + 1 < len(nums) and nums[j] == nums[j + 1]:
                    j += 1
            seen.add(nums[j])
            j += 1 

File: leetCode/JavaScript/15-3sum.js
//https://leetcode.com/problems/3sum/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const threeSum = (nums) => {
  if (nums && nums.length < 3) {
    return [];
  }
  nums.sort((a, b) => a - b);

  const res = [];

  for (let i = 0; i < nums.length - 2; i++) {
    let n1 = nums[i];
    if (n1 > 0) {
      break;
    }
    if (i - 1 >= 0 && n1 == nums[i - 1]) {
      continue;
    }

    let left = i + 1;
    let right = nums.length - 1;

    while (left < right) {
      let n2 = nums[left],
        n3 = nums[right];

      if (n1 + n2 + n3 === 0) {
        res.push([n1, n2, n3]);
        while (left < right && nums[left] == n2) {
          left++;
        }
        while (left < right && nums[right] == n3) {
          right--;
        }
      } else if (n1 + n2 + n3 < 0) {
        left++;
      } else {
        right--;
      }
    }
  }
  return res;
};
